pyiceberg/io/fsspec.py

```diff
@@ -194,13 +195,18 @@ def _gs(properties: Properties) -> AbstractFileSystem:
 def _adls(properties: Properties) -> AbstractFileSystem:
     from adlfs import AzureBlobFileSystem
 
-    for key, sas_token in {
-        key.replace(f"{ADLS_SAS_TOKEN}.", ""): value for key, value in properties.items() if key.startswith(ADLS_SAS_TOKEN)
-    }.items():
-        if ADLS_ACCOUNT_NAME not in properties:
-            properties[ADLS_ACCOUNT_NAME] = key.split(".")[0]
-        if ADLS_SAS_TOKEN not in properties:
-            properties[ADLS_SAS_TOKEN] = sas_token
+    # https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri#uri-syntax
+    if netloc := properties.get("netloc"):
+        account_uri = netloc.split("@")[-1]
+    else:
+        account_uri = None
+
+    if not properties.get(ADLS_ACCOUNT_NAME) and account_uri:
+        properties[ADLS_ACCOUNT_NAME] = account_uri.split(".")[0]
+
+    # Fixes https://github.com/apache/iceberg-python/issues/1146
+    if not properties.get(ADLS_SAS_TOKEN) and account_uri:
+        properties[ADLS_SAS_TOKEN] = properties.get(f"{ADLS_SAS_TOKEN}.{account_uri}")
```

Comment:
I'm a bit confused by this part. 

Given the uri `abfss://<file_system>@<account_name>.dfs.core.windows.net/<path>/<file_name>`, the `netloc` would be `<file_system>@<account_name>.dfs.core.windows.net`, and `account_uri` would be `<account_name>.dfs.core.windows.net` 

I presume we need to create a new `AzureBlobFileSystem` object for each unique netloc. We can continue to follow the pyarrow implementation here by providing both the `scheme` and `netloc`  to the FS initialization function
https://github.com/apache/iceberg-python/blob/80135451d030569259d83674ef147e0d6f62fd51/pyiceberg/io/pyarrow.py#L570